src/engine.rs

```diff
@@ -1,12 +1,10 @@
-use super::cas::Cas;
-use super::error::PipelineError;
+use crate::cas::Cas;
+use crate::error::Result;
 
 pub trait Engine {
-    fn process(&self, cas: &mut Cas) -> Result<(), PipelineError>;
+    fn process(&self, cas: &mut Cas) -> Result<()>;
 }
 
 pub trait Reader {
-    fn has_next(&mut self) -> bool;
-    fn execute(&mut self, cas: &mut Cas) -> Result<(), PipelineError>;
-    fn initialize(&mut self) -> Result<(), PipelineError>;
+    fn next_cas(&mut self) -> Option<Result<Cas>>;
```

Comment:
- Rather than asking if we have another and then giving it (which I know is like Java's iterators) we can just return an `Option`, which is how we do iterators in Rust.
- It seems like you just pass in a blank `Cas` to execute, so we should just return them instead.
- IMO, rather than having to call an initialize method on our reader, the reader should initialize itself upon creation. Look up RAII for more information on this pattern. It's used a lot in Rust, for example we have `File::open(path) -> Result<File>` instead of using `File::new(path) -> File` and then `File::open(&mut self) -> Result<()>`.